ext/standard/array.c

```diff
@@ -6620,23 +6620,32 @@ static enum php_array_find_result php_array_find(const HashTable *array, zend_fc
 		if (!str_key) {
 			ZVAL_LONG(&args[1], num_key);
 		} else {
+			ZEND_ASSUME(!HT_IS_PACKED(array));
 			ZVAL_STR(&args[1], str_key);
 		}
 
 		ZVAL_COPY_VALUE(&args[0], operand);
 
-		zend_result result = zend_call_function(&fci, &fci_cache);
+		zend_result result = zend_call_function(&fci, fci_cache);
 		ZEND_ASSERT(result == SUCCESS);
 
 		if (UNEXPECTED(Z_ISUNDEF(retval))) {
 			return PHP_ARRAY_FIND_EXCEPTION;
 		}
 
-		bool retval_true = zend_is_true(&retval);
-		zval_ptr_dtor(&retval);
-
-		/* This negates the condition, if negate_condition is true. Otherwise it does nothing with `retval_true`. */
-		retval_true ^= negate_condition;
+		bool retval_true;
+		switch (Z_TYPE(retval)) {
+			case IS_TRUE:
+				retval_true = !negate_condition;
+				break;
+			case IS_FALSE:
+				retval_true = negate_condition;
+				break;
+			default:
+				retval_true = zend_is_true(&retval) ^ negate_condition;
+				zval_ptr_dtor(&retval);
+				break;
+		}
```

Comment:
This can probably also apply to `array_filter()`. Perhaps this logic can be moved into some force-inlined:

```
	bool is_true(zval *retval, bool negate_condition) {
		switch (Z_TYPE(retval)) {
			case IS_TRUE:
				return !negate_condition;
			case IS_FALSE:
				return negate_condition;
			default:
				bool retval_true = zend_is_true(&retval) ^ negate_condition;
				zval_ptr_dtor(&retval);
				return retval_true;
		}
	}
```

that is then used for array_find and array_filter.